File: very hard/Shadows of the Knight - Episode 2/Python3/Shadows of the Knight - Episode 2.py
import sys

# Read building dimensions: width and height.
w, h = [int(i) for i in input().split()]
input()  # Number of jumps (not used in this solution)
x0, y0 = [int(i) for i in input().split()]  # Starting position

# x0, y0: previous position; x, y: current position.
x, y = x0, y0

# xs and ys represent the remaining possible windows where the bomb could be.
# Initially, all columns (xs) and all rows (ys) are possible.
xs, ys = range(w), range(h)

def narrow(x0, y0, x, y, xs, ys, info):
    """
    Narrow down the possible bomb location based on sensor feedback.
    
    Parameters:
        x0, y0 (int): Previous position.
        x, y (int): Current position.
        xs (range): Possible columns.
        ys (range): Possible rows.
        info (str): Sensor feedback ("WARMER", "COLDER", "SAME", "UNKNOWN").
    """
    print("narrow: x0={}, y0={}, x={}, y={}, info={}".format(x0, y0, x, y, info), file=sys.stderr)
    
    # Binary search on the x-axis (columns)
    if len(xs) != 1:
        if info == "UNKNOWN":  # First jump: no info available.
            pass
        elif info == "SAME":  # Distance remains the same: keep only equidistant positions.
            xs = [i for i in xs if abs(x0 - i) == abs(x - i)]
        elif info == "WARMER":  # Getting closer: keep positions that are closer.
            xs = [i for i in xs if abs(x0 - i) > abs(x - i)]
        else:  # COLDER: getting further: keep positions that are farther.
            xs = [i for i in xs if abs(x0 - i) < abs(x - i)]
    
    # Binary search on the y-axis (rows) when x-axis is narrowed to one value.
    else:
        if info == "UNKNOWN":
            pass
        elif info == "SAME":
            ys = [i for i in ys if abs(y0 - i) == abs(y - i)]
        elif info == "WARMER":
            ys = [i for i in ys if abs(y0 - i) > abs(y - i)]
        else:
            ys = [i for i in ys if abs(y0 - i) < abs(y - i)]
    
    print(xs, file=sys.stderr)  # Debug: show remaining columns.
    print(ys, file=sys.stderr)  # Debug: show remaining rows.
    return xs, ys

# Main game loop
while True:
    info = input()  # Sensor feedback: "WARMER", "COLDER", "SAME", or "UNKNOWN"
    
    # Narrow down the search area based on sensor feedback.
    xs, ys = narrow(x0, y0, x, y, xs, ys, info)
    
    # Update previous position.
    x0, y0 = x, y
    
    # Binary search on the x-axis (columns)
    if len(xs) != 1:
        # Bisection between x0 and x to cut the zone roughly in half.
        if x0 == 0 and len(xs) != w:
            x = (3 * xs[0] + xs[-1]) // 2 - x0
        elif x0 == w - 1 and len(xs) != w:
            x = (xs[0] + 3 * xs[-1]) // 2 - x0
        else:
            x = xs[0] + xs[-1] - x0
        
        # Avoid staying in the same position.
        if x == x0:
            x += 1
        x = min(max(x, 0), w - 1)
    
    # Transition to binary search on the y-axis.
    else:
        if x != xs[0]:
            x = x0 = xs[0]
            print(x, y)
            info = input()
        # Finalize: if only one row remains.
        if len(ys) == 1:
            y = ys[0]
        else:
            if y0 == 0 and len(ys) != h:
                y = (3 * ys[0] + ys[-1]) // 2 - y0
            elif y0 == h - 1 and len(ys) != h:
                y = (ys[0] + 3 * ys[-1]) // 2 - y0
            else:
                y = ys[0] + ys[-1] - y0
            y = min(max(y, 0), h - 1)
    
    # Output the new position.
    print(x, y)

File: very hard/Shadows of the Knight - Episode 2/TypeScript/Shadows of the Knight - Episode 2.ts
/**
 * /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 * // Function narrowDimension:                                                                                      //
 * //   Updates the search interval in one dimension                                                                  //
 * //   based on the feedback and the comparison between                                                              //
 * //   the previous position (x0 or y0) and the current position (x or y).                                           //
 * /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 */
function narrowDimension(prev: number, current: number, min: number, max: number, info: string): [number, number] {
    const sum = prev + current;

    if (info === "SAME") {
        if (sum % 2 === 0) {
            const middle = Math.floor(sum / 2);
            min = middle;
            max = middle;
        }
    } else if (info === "WARMER") {
        if (current > prev) {
            const newLower = Math.floor(sum / 2) + 1;
            if (newLower > min) min = newLower;
        } else if (current < prev) {
            const newUpper = sum % 2 === 0 ? Math.floor(sum / 2) - 1 : Math.floor(sum / 2);
            if (newUpper < max) max = newUpper;
        }
    } else if (info === "COLDER") {
        if (current > prev) {
            const newUpper = sum % 2 === 0 ? Math.floor(sum / 2) - 1 : Math.floor(sum / 2);
            if (newUpper < max) max = newUpper;
        } else if (current < prev) {
            const newLower = Math.floor(sum / 2) + 1;
            if (newLower > min) min = newLower;
        }
    }

    return [min, max];
}

/**
 * ////////////////////////////////////////////////////////////////
 * // Function narrow:                                           //
 * // Applies narrowDimension on x or y to update intervals      //
 * ////////////////////////////////////////////////////////////////
 */
function narrow(
    x0: number, y0: number, x: number, y: number,
    xMin: number, xMax: number, yMin: number, yMax: number, info: string
): [number, number, number, number] {
    console.error(`narrow: x0: ${x0} y0: ${y0} x_min: ${xMin} x_max: ${xMax} y_min: ${yMin} y_max: ${yMax} info: ${info}`);

    if (xMin !== xMax) {
        [xMin, xMax] = narrowDimension(x0, x, xMin, xMax, info);
    } else if (yMin !== yMax) {
        [yMin, yMax] = narrowDimension(y0, y, yMin, yMax, info);
    }

    console.error(`narrow : x_min: ${xMin} x_max: ${xMax} y_min: ${yMin} y_max: ${yMax}`);
    return [xMin, xMax, yMin, yMax];
}

// Initial input parsing
let inputs: string[] = readline().split(' ');
const w: number = parseInt(inputs[0]);
const h: number = parseInt(inputs[1]);
const _n: number = parseInt(readline()); // unused
inputs = readline().split(' ');
let x0: number = parseInt(inputs[0]);
let y0: number = parseInt(inputs[1]);
let x: number = x0;
let y: number = y0;

let xMin: number = 0;
let xMax: number = w - 1;
let yMin: number = 0;
let yMax: number = h - 1;

while (true) {
    let info: string = readline();

    [xMin, xMax, yMin, yMax] = narrow(x0, y0, x, y, xMin, xMax, yMin, yMax, info);

    console.error(`narrow results: x_min: ${xMin} x_max: ${xMax} y_min: ${yMin} y_max: ${yMax}`);

    x0 = x;
    y0 = y;

    if (xMin !== xMax) {
        if (x0 === 0 && xMax + xMin + 1 !== w) {
            x = Math.floor((3 * xMin + xMax) / 2) - x0;
        } else if (x0 === w - 1 && xMax + xMin + 1 !== w) {
            x = Math.floor((xMin + 3 * xMax) / 2) - x0;
        } else {
            x = xMax + xMin - x0;
        }

        if (x === x0) x++;
        x = Math.max(0, Math.min(w - 1, x));
    } else {
        if (x !== xMin) {
            x = xMin;
            x0 = x;
            console.log(`${x} ${y}`);
            info = readline();
        }

        if (yMin !== yMax) {
            if (y0 === 0 && yMax + yMin + 1 !== h) {
                y = Math.floor((3 * yMin + yMax) / 2) - y0;
            } else if (y0 === h - 1 && yMax + yMin + 1 !== h) {
                y = Math.floor((yMin + 3 * yMax) / 2) - y0;
            } else {
                y = yMax + yMin - y0;
            }

            if (y === y0) y++;
            y = Math.max(0, Math.min(h - 1, y));
        } else {
            y = yMin;
        }
    }

    console.log(`${x} ${y}`);
}
